File: src/0374_guess_number_higher_or_lower.js
const PICK = 6;
function guess(num) {
    if (num === PICK) return 0;
    if (num < PICK) return 1;
    return -1;
}
function guessNumber(n) {
    let low = 1;
    let high = n;
    while (low <= high) {
        const mid = low + Math.floor((high - low) / 2);
        const myGuess = guess(mid);
        if (myGuess === 1) low = mid + 1;
        else if (myGuess === -1) high = mid - 1;
        else return mid;
    }
    return -1;
}

module.exports = { guessNumber };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(guessNumber(10), 6);
    assert.deepStrictEqual(guessNumber(6), 6);
}

File: src/0374_guess_number_higher_or_lower.rs
struct Solution;

const PICK: i32 = 6;

fn guess(num: i32) -> i32 {
    if num == PICK {
        0
    } else if num < PICK {
        1
    } else {
        -1
    }
}

impl Solution {
    pub fn guess_number(n: i32) -> i32 {
        let mut low = 1;
        let mut high = n;

        while low <= high {
            let mid = low + (high - low) / 2;
            let my_guess = guess(mid);
            if my_guess == 1 {
                low = mid + 1;
            } else if my_guess == -1 {
                high = mid - 1;
            } else {
                return mid;
            }
        }
        -1
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(Solution::guess_number(10), 6);
        assert_eq!(Solution::guess_number(6), 6);
    }
}
